File: backend/src/role/role.service.js
const Role = require('./role.model'); // Importing the Role model

// Function to create a new role
async function createRole(roleId, roleName, accessList, status) {
    try {
        const role = new Role({
            roleId,
            roleName,
            accessList,
            status,
        });
        await role.save();
        return role;
    } catch (error) {
        throw error;
    }
}

// Function to find a role by _id
async function findRoleById(id) {
    try {
        const role = await Role.findById(id);
        return role;
    } catch (error) {
        throw error;
    }
}

// Function to find a role by roleId
async function findRoleByRoleId(roleId) {
    try {
        const role = await Role.findOne({ roleId });
        return role;
    } catch (error) {
        throw error;
    }
}

// Function to find all roles
async function findAllRoles() {
    try {
        const roles = await Role.find();
        return roles;
    } catch (error) {
        throw error;
    }
}

// Function to update role information
async function updateRole(roleId, updates) {
    try {
        // Check if the roleId is being updated
        if (updates.roleId) {
            // Check if the new roleId already exists
            const existingRole = await Role.findOne({ roleId: updates.roleId });
            if (existingRole && existingRole.roleId !== roleId) {
                throw new Error("Role with this roleId already exists.");
            }
        }
        // Update the role information
        const role = await Role.findOneAndUpdate({ roleId }, updates, { new: true });
        return role;
    } catch (error) {
        throw error;
    }
}

// Function to delete a role
async function deleteRole(roleId) {
    try {
        const role = await Role.findOneAndDelete({ roleId });
        return role;
    } catch (error) {
        throw error;
    }
}

module.exports = {
    createRole,
    findRoleById,
    findAllRoles,
    findRoleByRoleId,
    updateRole,
    deleteRole,
};

File: QuestCraftAPI/src/main/java/org/example/questcraftapi/role/RoleService.java
package org.example.questcraftapi.role;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class RoleService {

    @Autowired
    private RoleRepository roleRepository;

    public RoleDocument createRole(RoleDocument role) {
        return roleRepository.save(role);
    }

    public RoleDocument findRoleById(String id) {
        return roleRepository.findById(id).orElseThrow();
    }

    public RoleDocument findRoleByRoleId(String roleId) {
        return roleRepository.findByRoleId(roleId);
    }

    public List<String> findAllActiveRoleIds() {
        return roleRepository.findAll().stream()
                .filter(role -> "active".equals(role.getStatus()))
                .map(RoleDocument::getRoleId)
                .collect(Collectors.toList());
    }

    public List<RoleDocument> findAllRoles() {
        return roleRepository.findAll();
    }

    public RoleDocument updateRole(String roleId, RoleDocument updates) {
        RoleDocument existingRole = roleRepository.findByRoleId(roleId);
        if (existingRole == null) {
            throw new RuntimeException("Role not found.");
        }
        // Update the role information
        existingRole.setAccessList(updates.getAccessList());
        existingRole.setStatus(updates.getStatus());
        return roleRepository.save(existingRole);
    }

    public void deleteRole(String roleId) {
        RoleDocument role = roleRepository.findByRoleId(roleId);
        if (role == null) {
            throw new RuntimeException("Role not found.");
        }
        roleRepository.delete(role);
    }
}